### PR TITLE
nixos/syncthing: add versioning

### PR DESCRIPTION
**Pull request by mrVanDalo**
###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @FRidh

### CODE REVIEW

File: nixos/modules/services/networking/syncthing.nix [hunk 1]
```diff
@@ -220,6 +221,70 @@ in {
                 '';
               };
 
+
+              versioning = mkOption {
+                default = null;
+                description = ''
+                  how to keep changed/deleted files with syncthing.
```
Comment (FRidh) [inline nixos/modules/services/networking/syncthing.nix]: Capital letter to start description

**Comment by FRidh** [comment]
Note you can test whether the issue still occurs with
```
$ nix-build -A metrics pkgs/top-level/release.nix 
```

**Comment by mrVanDalo** [comment]
> Note you can test whether the issue still occurs with
> 
> ```
> $ nix-build -A metrics pkgs/top-level/release.nix 
> ```

which issue?

**Comment by FRidh** [comment]
the lack of descriptions broke the metrics job, preventing a channel from advancing

**Comment by FRidh** [comment]
Basically, when changing a module, it's good not to just test the module (which I am sure you did), but also this job.

**Comment by mrVanDalo** [comment]
I see, when I have some more time (I guess on Monday or Sunday), I will have a look at this metrics job. Never heard of it before, also @GrahamcOfBorg  did not post anything. Is this party of https://hydra.nixos.org ? I also couldn't find any information about this in the https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md.

Btw I think the pull request can be merged now.

**Comment by FRidh** [comment]
ofborg does not evaluate NixOS tests unfortunately. https://github.com/NixOS/ofborg/issues/315
The metrics job is a channel requirement https://hydra.nixos.org/job/nixpkgs/trunk/unstable#tabs-constituents

**Comment by mrVanDalo** [comment]
@FRidh thank your explanation and your links. When I ran the command, you posted, in the branch of my changes, I get this error : 

```
running nix-env -f /nix/store/xq9xlwifn82gvvkqbh6xam7sfyrx7b04-nixpkgs -qa
error: executing '': No such file or directory
{
  "cpuTime": 6.69466,
  "envs": {
    "number": 2508238,
    "elements": 6373342,
    "bytes": 91118544
  },
  "list": {
    "elements": 2231186,
    "bytes": 17849488,
    "concats": 147348
  },
  "values": {
    "number": 8350889,
    "bytes": 200421336
  },
  "symbols": {
    "number": 308128,
    "bytes": 13451059
  },
  "sets": {
    "number": 1506164,
    "bytes": 669575824,
    "elements": 27396938
  },
  "sizes": {
    "Env": 16,
    "Value": 24,
    "Bindings": 8,
    "Attr": 24
  },
  "nrOpUpdates": 897367,
  "nrOpUpdateValuesCopied": 21716914,
  "nrThunks": 6835863,
  "nrAvoided": 3856394,
  "nrLookups": 1944563,
  "nrPrimOpCalls": 895598,
  "nrFunctionCalls": 2133214,
  "gc": {
    "heapSize": 990572544,
    "totalBytes": 1272965408
  }
}
6.85user 0.64system 0:07.49elapsed 99%CPU (0avgtext+0avgdata 1448796maxresident)k
0inputs+72outputs (0major+365520minor)pagefaults 0swaps

parse error: Invalid numeric literal at line 1, column 6
builder for '/nix/store/3y9azq795wf3q99cv2zmm3szzhx6jgfp-nixpkgs-metrics.drv' failed with exit code 4
error: build of '/nix/store/3y9azq795wf3q99cv2zmm3szzhx6jgfp-nixpkgs-metrics.drv' failed
```

**Comment by FRidh** [comment]
I noticed it as well, even though the test passes on Hydra! Will need to investigate this further.

**Review by Lassulus — APPROVED** [review]
so lets try this again?

**Comment by mrVanDalo** [comment]
looks like it worked. https://hydra.nixos.org/job/nixpkgs/trunk/metrics

**Comment by contrun** [comment]
It seems that syncthing is now using the following format

```
        <versioning type="staggered">
            <param key="cleanInterval" val="3600"></param>
            <param key="maxAge" val="31536000"></param>
            <cleanupIntervalS>3600</cleanupIntervalS>
            <fsPath>~/.cache/syncthing_versioning/calibre</fsPath>
            <fsType>basic</fsType>
        </versioning>
```

This stanza is generated from GUI. The GUI generates xml like this by pull request https://github.com/syncthing/syncthing/pull/7407 .

**Pull request merged**